### ISSUE TITLE
Brian Tutorial second attempt!

### ISSUE THREAD

**Issue by rcaze** [brian-team/brian2#632]
Dear all, 

I am working on the Brian tutorial and instead of using the HH to see spike I would like to use the hard threshold method, as it will be faster (for network simulation!).

The thresholding is already implemented in Brian and what I would like to do is to set a height for the generated spike (currently it is set to 0). For instance I would like to set the threshold to -50mV and create a spike that goes to 70mV and reset to -75mV almost immediatly (next time step). I would love to have an optional argument in `SpatialNeuron` that could be called `spike_height`.

This feature would not only be nice for plotting (to have beautiful spikes) but also for bAP that would backpropagate into the dendritic tree.

Are all the elements already there in Brian? If no I am happy to fork and work on this feature. Where should I start to look? I looked a bit at the code and `SpatialNeuron` refers to `NeuronGroup` for the thresholding thing.

**Comment by mstimberg** [comment]
> I am working on the Brian tutorial and instead of using the HH to see spike I would like to use the hard threshold method, as it will be faster (for network simulation!).

You want to use this in a `SpatialNeuron` model? You can of course have a multicompartmental  integrate-and-fire model, but there are quite a few questions to answer: do you have a threshold in each compartment, or just one in the soma? What happens at a spike? In a standard I&F model, you'd just register the spike and reset the membrane potential, the voltage never goes up. How do you make the spike propagate along the axon then?

And finally, I hope you are aware that you can have several `SpatialNeuron` objects to simulate a small network, but this is very inefficient -- there is no such thing as a `SpatialNeuronGroup`.

> The thresholding is already implemented in Brian and what I would like to do is to set a height for the generated spike (currently it is set to 0).

As I described above, there is no such thing as a "generated spike" in an I&F model, and Brian does not set anything to 0.

> For instance I would like to set the threshold to -50mV and create a spike that goes to 70mV and reset to -75mV almost immediatly (next time step).

There are (somewhat complicated) ways to achieve that, but it would not be a well-defined model (its behaviour would change with the length of a time step).

Well, in sum, I'd stick with a HH model if you want to write a tutorial for `SpatialNeuron` :)

**Comment by rcaze** [comment]
> You want to use this in a SpatialNeuron model? You can of course have a multicompartmental integrate-and-fire model, but there are quite a few questions to answer: do you have a threshold in each compartment, or just one in the soma? What happens at a spike? In a standard I&F model, you'd just register the spike and reset the membrane potential, the voltage never goes up. How do you make the spike propagate along the axon then?

In the code below and I am adding the spikes post-hoc (that is a dirty trick). I want the spike to back propagate, one way to do that would be to set the voltage of the soma to 20 mV when the threshold is crossed (this is what I have done in the video with the bioarxiv article)  

> And finally, I hope you are aware that you can have several SpatialNeuron objects to simulate a small network, but this is very inefficient -- there is no such thing as a SpatialNeuronGroup.

This is exactly what I want to create. I am sure it can be turned into an efficient process.

> As I described above, there is no such thing as a "generated spike" in an I&F model, and Brian does not set anything to 0.

Another solution would be to use what is presented in this example: http://brian2.readthedocs.org/en/2.0b4/examples/frompapers.Brette_Gerstner_2005.html

    For instance I would like to set the threshold to -50mV and create a spike that goes to 70mV and reset to -75mV almost immediatly (next time step).

> There are (somewhat complicated) ways to achieve that, but it would not be a well-defined model (its behaviour would change with the length of a time step).

> Well, in sum, I'd stick with a HH model if you want to write a tutorial for SpatialNeuron :)

Simulating the HH equation is not only lenghty but also unrealistic! I am more than ready to work hard on making network of neuron more complex than point neurons

see code below it produces the attached fig
```python
# Brian tutorial (for NEURON lovers)
import matplotlib.pyplot as plt
import brian2 as br2
from brian2 import siemens, um, mV, cm, uF, ohm, ms, namp

# Equation to create a leak current
EL = -65*mV
gL = 0.0003*siemens/cm**2
pas = '''
Im = gL * (EL - v) : amp/meter**2
I : amp (point current)
'''

# Create the geometrical objects corresponding to the two compartments
e_neuron = br2.Soma(diameter=10*um)

# Insert biophysical mechanism in the morphology object
neuron = br2.SpatialNeuron(morphology=e_neuron, model=pas,
                           method="exponential_euler",
                           Cm=1*uF/cm**2, Ri=123*ohm*cm,
                           threshold="v>-40*mV",
                           reset="v=-65*mV")

# Set a state monitor to look at the voltage
M = br2.StateMonitor(neuron, 'v', record=True)

# Run the simulation
neuron.v = -65*mV
br2.run(45*ms)
neuron.I = 0.05*namp
br2.run(10*ms)
neuron.I = 0*namp
br2.run(45*ms)

# Add spikes to the recorded somatic voltage trace
# Conver because the Brian array do not support fancy indexing (it seems)
soma_v = br2.asarray(M.v[0])*1e3
soma_v[soma_v > -41] = 20

# Plot the result
plt.plot(M.t/ms, soma_v)
plt.xlabel('Time (ms)')
plt.ylabel('v')
plt.ylim(-80, 50)
plt.savefig("soma_spike.png")
```

This code takes an awful long time to run (something like 3s). Why?
![soma_spike](https://cloud.githubusercontent.com/assets/4049489/12490003/31267548-c06b-11e5-8646-4149a92af4a4.png)

**Comment by rcaze** [comment]
This code is producing what I do *not* want. 

```python
# Brian tutorial (for NEURON lovers)
import matplotlib.pyplot as plt
import brian2 as br2
from brian2 import siemens, um, mV, cm, uF, ohm, ms, namp

# Equation to create a leak current
EL = -65*mV
gL = 0.0003*siemens/cm**2
pas = '''
Im = gL * (EL - v) : amp/meter**2
I : amp (point current)
'''

# Create the geometrical objects corresponding to the two compartments
e_neuron = br2.Soma(diameter=10*um)
e_neuron.dend = br2.Cylinder(diameter=1*um,
                             x=0*um, y=200*um, z=0*um,
                             n=5)
# Create attribut used optionally
e_neuron.dend.name = "dend"
e_neuron.sections = {"soma": e_neuron, "dend": e_neuron.dend}

neuron = br2.SpatialNeuron(morphology=e_neuron, model=pas,
                           method="exponential_euler",
                           Cm=1*uF/cm**2, Ri=123*ohm*cm,
                           threshold="v>-40*mV",
                           threshold_location=0,
                           reset="v=-65*mV")


neuron.v = -65*mV
M = br2.StateMonitor(neuron, 'v', record=True)

# Run simulation
br2.run(45*ms)
neuron.I = 0.02*namp
br2.run(10*ms)
neuron.I = 0*namp
br2.run(45*ms)

# Add spikes to the recorded somatic voltage trace
# Conver because the Brian array do not support fancy indexing (it seems)
soma_v = br2.asarray(M.v[0])*1e3
#soma_v[soma_v > -41] = 10
soma_v[soma_v > -41] = 20


plt.plot(M.t/ms, soma_v, label="soma")
plt.plot(M.t/ms, M.v[5]/mV, label="dend(1)")
plt.xlabel('Time (ms)')
plt.ylabel('v')
plt.ylim(-80, 50)
plt.legend()
plt.savefig("soma_dend_spike.png")
```
Obviously and as expected the spikes does not backpropagate. But we are close.. 
![soma_dend_spike](https://cloud.githubusercontent.com/assets/4049489/12490650/86ec25ba-c06e-11e5-8a92-0fc01e670038.png)

**Comment by mstimberg** [comment]
I have to run, but just very quickly:

> In the code below and I am adding the spikes post-hoc (that is a dirty trick).

I don't see this as a dirty trick, this is what you usually do in I&F models for plotting.

> I want the spike to back propagate, one way to do that would be to set the voltage of the soma to 20 mV when the threshold is crossed (this is what I have done in the video with the bioarxiv article)

I still don't think this is a good way to do it, because it means that the effect of a spike scales with `dt`. If the neighbouring compartments propagate the spike, then it does not matter, but for passive compartments that would change the effect of the spike.

I don't think this changes fundamentally if you use the EIF model instead of a simpler LIF.

> Simulating the HH equation is not only lenghty but also unrealistic!

? The HH equations are not "realistic" for a single compartment neuron, but for a multi-compartment they are certainly fine, no? 

Work on a `SpatialNeuronGroup` would certainly be appreciated, but this is a non-trivial task that involves quite a bit of work and decisions for syntax etc. This would rather be a long-term project, i.e. something that will take a few months until done (starting on this should also wait until we finished our current changes to the `SpatialNeuron` class).

**Comment by rcaze** [comment]
It would be cool to have something in the optional argument of SpatialNeuron like `reset="v[-1]=20*mV and v=-65*mV"` becaue the dendritic compartments "feels" the reset.

**Comment by rcaze** [comment]
;} Not afraid of a long term goal. Working on network of spatial neuron was in the first version of my PhD, I was told "What's the point? Everybody knows that neurons are simple, yeah?", this is how I started to work on dendrites and single neuron computation. It was already 7 years ago ;). And it seems that I found some help with a great software.

**Comment by rcaze** [comment]
It seems that the `event` optional argument in `NeuroGroup` might do the job. I kinda understand your point about dt. How is the voltage dealt during `reset`? (the hyperpolarization seems to propagate)

**Comment by rcaze** [comment]
Alternatively, would it be possible to have different equations depending on the group (section). For instance inserting some equation in the soma and another in the dendrites?

**Comment by mstimberg** [comment]
There is currently no way to have compartment-specific equations or to access individual compartments in code strings (like the reset you mentioned above). The standard workaround is to have a condition and multiply things with 0 or 1 depending on the condition. E.g. this would use a different reset for the compartment with the index 0:
```Python
reset='int(i == 0) * reset_0 + int(i != 0) * reset_other'
```

**Comment by rcaze** [comment]
I mean that the `reset` seems to backpropagate. The `event` in `NeuronGroup` seems a way to generate a spike. The event would be set the voltage to 20mV. a char like `v = 20*mV`

**Comment by mstimberg** [comment]
> I mean that the reset seems to backpropagate.

The reset only sets the compartment which was above threshold but since the compartments are coupled, it is possible that more than one compartment goes above threshold before the threshold condition is checked and consequently, more than one compartment is reset. Have a look at https://brian2.readthedocs.org/en/latest/advanced/scheduling.html to see how one time step is organized (a `StateMonitor` can record at any point during the time step by specifying the `when` argument, this can be useful for diagnostics). 

> The event in NeuronGroup seems a way to generate a spike. The event would be set the voltage to 20mV. a char like v = 20*mV

An event defines a *condition*, i.e. it is like a threshold (in fact, `threshold=condition` is equivalent to `events={'spike': condition}`). You can trigger code based on non-standard events with the `run_on_event` method. But sure, you can combine all this to get a "spike" for a single time step, something along the lines of
```Python
eqs = '''Im = ...
            ...
            was_spiking : boolean
         '''
neuron = SpatialNeuron(..., eqs, threshold='v>v_th and not(was_spiking)',
                       reset='v = v_spike; was_spiking = True',
                       events={'after_spike_reset': 'was_spiking'}, ...)
neuron.run_on_event('after_spike_reset', 'v = v_reset; was_spiking = False')
```
Given the scheduling (see link above), at the time step where the membrane potential crosses the threshold, the membrane potential is set to a high value and `was_spiking` is set. The next time step, the high membrane potential will trigger strong current flow to the neighbouring compartments and after the integration of the membrane potential, the membrane potential of the spiking compartment will be reset to the low value and `was_spiking` is reset.
The sudden jumps in the membrane potential might not be great for numerical stability, though.

**Comment by rcaze** [comment]
It seems that we have a solution here! Good news for weekend. Will (help) work on an implementation of it next week.

**Comment by rcaze** [comment]
About including diffferent mechanism in each compartment. There can be, however, different value of a variable in each compartment. For instance one might define a `pas` variable that would be 0 for certain compartment (no leak there) and 1 for others (where to insert pas).

**Comment by rcaze** [comment]
I cannot find the `events` nor the `run_on_event` method in `SpatialNeuron`

**Comment by mstimberg** [comment]
> I cannot find the events nor the run_on_event method in SpatialNeuron

Hmm, it seems we forgot to add the `events` keyword to the `SpatialNeuron` initializer when we added it to `NeuronGroup`. The `run_one_event` method is just inherited from `NeuronGroup`. As a workaround for now, you should be able to add a new event like this (I did not test this, though):
```Python
from brian2.groups.neurongroup import Thresholder
...
neuron = SpatialNeuron(...)
neuron.events['my_event'] = my_event_condition
thresholder = Thresholder(neuron, 'my_event')
neuron.thresholder['my_event'] = thresholder
neuron.contained_objects.append(thresholder)
```

**Comment by rcaze** [comment]
Thank you, I am going to try that. I saw in the Brian code that there was no `events` keywords. I also saw that there was a lot of `bla=bla` to pass keywords arg around. I had a lot of that in my code too. As a workaround I now use this snippet.

```python
# Demonstrate how to pass arguments from function to function
import inspect


def bar(arg, karg=1):
    # Function to which we are going to pass the arguments
    return arg + karg


def foo(arg, karg=2):
    # Use inspect to extract all arguments from foo
    frame = inspect.currentframe()
    args, _, _, values = inspect.getargvalues(frame)
    # Build a dictionary to pass to bar
    to_pass = {arg: values[arg] for arg in args}
    return bar(**to_pass)

if __name__ == "__main__":
    print foo(1, karg=3)
```

**Comment by rcaze** [comment]
Python throws me an error when I try `run_on_event`

**Comment by mstimberg** [comment]
> Python throws me an error when I try run_on_event

It might be helpful to know what error...

But I realized that I forgot something in my workaround above (has to be added before the creation of the `Thresholder` object):
```Python
neuron.variables.add_array('_{}space'.format('my_event'), unit=1,
                           size=neuron.N+1, dtype=np.int32)
```

**Comment by rcaze** [comment]
An AttributeError: No attribute with name run_on_event :)

**Comment by rcaze** [comment]
I used this 

```
python
neuron.events = {}
```
less black magic but maybe incorrect

**Comment by mstimberg** [comment]
About the `AttributeError`: this means you have an outdated version of Brian 2, `run_on_event` has been introduced in the 2.0b4 version.

I am not sure what `neuron.events = {}` is supposed to do, if you don't define any events, that's already the values of `neuron.events` anyway (but then you cannot use `run_on_event` of course).

**Comment by rcaze** [comment]
Almost there, my neuron spikes but does not reset. I included the following code

```python

neuron = br2.SpatialNeuron(morphology=e_neuron, model=pas,
                           method="exponential_euler",
                           Cm=1*uF/cm**2, Ri=123*ohm*cm,
                           threshold="v>-40*mV and not(was_spiking)",
                           threshold_location=0,
                           reset="v= 20*mV; was_spiking = True")

event_name = 'reset_aft_spike'
neuron.events[event_name] = 'was_spiking'
neuron.variables.add_array('_{}space'.format(event_name), unit=1,
                           size=neuron.N+1, dtype=np.int32)
thresholder = Thresholder(neuron, 'end')
neuron.thresholder[event_name] = thresholder
neuron.contained_objects.append(thresholder)
neuron.run_on_event(event_name, 'v = -65*mV; was_spiking = False')
```

But I obtain only the following thing. The event is not run, what am I doing wrong?
![soma_dend_spike](https://cloud.githubusercontent.com/assets/4049489/12587450/4e86f344-c44c-11e5-9662-bd195b9b1001.png)

**Comment by rcaze** [comment]
By the way the jump in dendritic voltage shows that the spike does propagate!

**Comment by mstimberg** [comment]
> But I obtain only the following thing. The event is not run, what am I doing wrong?

You'll have to tell the `Thresholder` to use the event you specified (by default it uses the `spike` event), and you probably don't want it to be checked in the `end` slot, but rather in the default `threshold` slot (otherwise you'd already reset the membrane potential at the end of the time step where it triggered the spike, i.e. the high membrane potential would never be "seen" by the other compartments):
```Python
thresholder = Thresholder(neuron, event=event_name)
```

**Comment by rcaze** [comment]
Thank you the ```event=``` optional argument was the missing piece! (I should have guessed because the thresholder was not used properly otherwise. I now have a spiking neuron! Do you plan to put the ```event=``` in a future release? 
![soma_bap_spike](https://cloud.githubusercontent.com/assets/4049489/12588371/47b61dac-c450-11e5-91d0-de93bf396c3b.png)

**Comment by mstimberg** [comment]
> Do you plan to put the event= in a future release?

Yes, we just forgot to include it. It should be a trivial fix, I opened an issue about it here: #637
